### PR TITLE
✨ Source S3: Add handling NoSuchBucket error

### PR DESCRIPTION
**Pull request by tolik0**
## What
Add raising config error in case of wrong bucket name.
Closes:
https://github.com/airbytehq/airbyte/issues/25835

## How
Added a new type of error to file-based cdk to avoid catching in file-based cdk and just raise AirbyteTracedException config error.

### CODE REVIEW

**Comment by airbyte-oss-build-runner** [comment]
##  source-s3 test report (commit [`75069a5c82`](https://github.com/airbytehq/airbyte/pull/31383/commits/75069a5c82061f589353fb6e1627b2d5698946b3)) - ❌

⏲️  Total pipeline duration: 28.60s 

| Step                                                      | Result   |
|:----------------------------------------------------------|:---------|
| Build source-s3 docker image for platform(s) linux/x86_64 | ❌        |
| Code format checks                                        | ✅        |
| Validate metadata for source-s3                           | ✅        |
| Connector version semver check                            | ✅        |
| Connector version increment check                         | ✅        |
| QA checks                                                 | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/tolik0_source-s3_wrong-bucket-config-error/1697200427/75069a5c82061f589353fb6e1627b2d5698946b3/source-s3/4.2.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:75069a5c82061f589353fb6e1627b2d5698946b3)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```

File: airbyte-cdk/python/airbyte_cdk/sources/file_based/availability_strategy/default_file_based_availability_strategy.py [hunk 1]
```diff
@@ -64,6 +64,8 @@ def check_availability_and_parsability(
     def _check_list_files(self, stream: "AbstractFileBasedStream") -> List[RemoteFile]:
         try:
             files = stream.list_files()
+        except CustomFileBasedSourceException as custom_exc:
```
Comment (clnoll) [inline airbyte-cdk/python/airbyte_cdk/sources/file_based/availability_strategy/default_file_based_availability_strategy.py:67]: @tolik0 thanks for this PR, I agree that we want to be able customize the exceptions.

Looking at this code: the error that's relevant to the user is still that there was a problem listing files, so I think we want to keep the `CheckAvailabilityError`, but want to make sure we can customize the message, e.g. for things like missing bucket.

Since we're already raising `CheckAvailabilityError ... from exc`, we should get the missing bucket error as part of the stacktrace when we raise the `CheckAvailabilityError`, right?

If that's obscuring the error too much, instead of raising  `CustomFileBasedSourceException`, WDYT about giving it a "message" attribute, then here we'll raise `CheckAvailabilityError(custom_exc.message, stream=stream.name) from exc`?

Comment (tolik0) [reply airbyte-cdk/python/airbyte_cdk/sources/file_based/availability_strategy/default_file_based_availability_strategy.py]: @clnoll Thanks for feedback!
I've removed **CustomFileBasedSourceException**. Do you think this is better? Should I split this PR into two: one for file-based CDK and another for S3 changes?

File: airbyte-cdk/python/airbyte_cdk/sources/file_based/availability_strategy/default_file_based_availability_strategy.py [hunk 2]
```diff
@@ -71,7 +71,8 @@ def _check_list_files(self, stream: "AbstractFileBasedStream") -> List[RemoteFil
         try:
             files = stream.list_files()
         except Exception as exc:
-            raise CheckAvailabilityError(FileBasedSourceError.ERROR_LISTING_FILES, stream=stream.name) from exc
```
Comment (clnoll) [inline airbyte-cdk/python/airbyte_cdk/sources/file_based/availability_strategy/default_file_based_availability_strategy.py:74]: I think what I'd do here is continue to raise the CustomException from your previous version, but catch it and re-raise with a CheckAvailabilityError, like this:

```
try:
    ...
except CustomFileBasedException as exc:
    raise CheckAvailabilityError(exc.error_message, stream=stream.name) from exc
except Exception as exc:
    ...
```

This way we can avoid having to do the `hasattr` check. Plus, I think we'll probably want to use your CustomFileBasedException in other places too as the file-based CDK matures.

File: airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py [hunk 3]
```diff
@@ -68,25 +69,30 @@ def get_matching_files(self, globs: List[str], prefix: Optional[str], logger: lo
...
 
         try:
-            if prefixes:
-                for prefix in prefixes:
-                    for remote_file in self._page(s3, globs, self.config.bucket, prefix, seen, logger):
-                        total_n_keys += 1
-                        yield remote_file
-            else:
-                for remote_file in self._page(s3, globs, self.config.bucket, None, seen, logger):
+            for current_prefix in (prefixes if prefixes else [None]):
```
Comment (clnoll) [inline airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py]: Why is this change needed?

File: airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py [hunk 4]
```diff
@@ -68,25 +69,30 @@ def get_matching_files(self, globs: List[str], prefix: Optional[str], logger: lo
...
-            raise ErrorListingFiles(
-                FileBasedSourceError.ERROR_LISTING_FILES,
-                source="s3",
-                bucket=self.config.bucket,
-                globs=globs,
-                endpoint=self.config.endpoint,
-            ) from exc
+            self.raise_error_listing_files(FileBasedSourceError.ERROR_LISTING_FILES, globs, exc)
+
+    def raise_error_listing_files(self, error_message: Union[str, FileBasedSourceError], globs: List[str], exc: Optional[Exception] = None):
```
Comment (clnoll) [inline airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py:87]: Why do we need a helper method for this?

**Review by clnoll — COMMENTED** [review]
Thanks @tolik0! I think I'm actually preferring to keep the `CustomFileBasedSourceException`. I left an example of what I have in mind in a comment.

I think you will have to split this into 2 PRs since the new CDK version won't be available yet when the source-s3 connector is published.

**Comment by airbyte-oss-build-runner** [comment]
##  source-s3 test report (commit [`c60529b0df`](https://github.com/airbytehq/airbyte/pull/31383/commits/c60529b0df11c5a6a2c83e8bd3810e3be3ed2faf)) - ❌

⏲️  Total pipeline duration: 16mn30s 

| Step                                                      | Result   |
|:----------------------------------------------------------|:---------|
| Build source-s3 docker image for platform(s) linux/x86_64 | ✅       |
| Unit tests                                                | ✅       |
| Acceptance tests                                          | ❌       |
| Code format checks                                        | ✅       |
| Validate metadata for source-s3                           | ✅       |
| Connector version semver check                            | ✅       |
| Connector version increment check                         | ✅       |
| QA checks                                                 | ✅       |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/tolik0_source-s3_wrong-bucket-config-error/1697668856/c60529b0df11c5a6a2c83e8bd3810e3be3ed2faf/source-s3/4.1.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:c60529b0df11c5a6a2c83e8bd3810e3be3ed2faf)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```

File: airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py (hunk 3 above)
Comment (tolik0) [reply airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py]: I simplified the code to avoid duplication

File: airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py (hunk 4 above)
Comment (tolik0) [reply airbyte-integrations/connectors/source-s3/source_s3/v4/stream_reader.py:87]: I used the helper method to reduce repetitive error-raising code.

File: airbyte-cdk/python/airbyte_cdk/sources/file_based/availability_strategy/default_file_based_availability_strategy.py (hunk 2 above)
Comment (tolik0) [reply airbyte-cdk/python/airbyte_cdk/sources/file_based/availability_strategy/default_file_based_availability_strategy.py:74]: I've updated the code to use CustomFileBasedException and re-raise as CheckAvailabilityError, removing the hasattr check. Thanks for the suggestion.

**Comment by airbyte-oss-build-runner** [comment]
##  source-s3 test report (commit [`3c18702032`](https://github.com/airbytehq/airbyte/pull/31383/commits/3c18702032a6772ab24715e72dd8245dfe39e028)) - ❌

⏲️  Total pipeline duration: 02mn36s 

| Step                                                      | Result   |
|:----------------------------------------------------------|:---------|
| Build source-s3 docker image for platform(s) linux/x86_64 | ❌       |
| Code format checks                                        | ❌       |
| Validate metadata for source-s3                           | ✅       |
| Connector version semver check                            | ✅       |
| Connector version increment check                         | ✅       |
| QA checks                                                 | ✅       |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/tolik0_source-s3_wrong-bucket-config-error/1697673618/3c18702032a6772ab24715e72dd8245dfe39e028/source-s3/4.1.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:3c18702032a6772ab24715e72dd8245dfe39e028)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-s3 test report (commit [`30e354ac81`](https://github.com/airbytehq/airbyte/pull/31383/commits/30e354ac814f270ad1e91c33dfc994851369e670)) - ❌

⏲️  Total pipeline duration: 14.54s 

| Step                                                      | Result   |
|:----------------------------------------------------------|:---------|
| Build source-s3 docker image for platform(s) linux/x86_64 | ❌       |
| Code format checks                                        | ✅       |
| Validate metadata for source-s3                           | ✅       |
| Connector version semver check                            | ✅       |
| Connector version increment check                         | ✅       |
| QA checks                                                 | ✅       |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/tolik0_source-s3_wrong-bucket-config-error/1697674485/30e354ac814f270ad1e91c33dfc994851369e670/source-s3/4.1.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:30e354ac814f270ad1e91c33dfc994851369e670)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```

**Review by clnoll — APPROVED** [review]
Thanks @tolik0, looks great!

**Review by bazarnov — APPROVED** [review]

**Comment by airbyte-oss-build-runner** [comment]
##  source-s3 test report (commit [`3edc8af363`](https://github.com/airbytehq/airbyte/pull/31383/commits/3edc8af363b692abdbacba2141a26691a03ea52e)) - ❌

⏲️  Total pipeline duration: 03mn22s 

| Step                                                     | Result   |
|:---------------------------------------------------------|:---------|
| Build source-s3 docker image for platform(s) linux/amd64 | ❌       |
| Code format checks                                       | ❌       |
| Validate metadata for source-s3                          | ✅       |
| Connector version semver check                           | ✅       |
| Connector version increment check                        | ✅       |
| QA checks                                                | ✅       |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/tolik0_source-s3_wrong-bucket-config-error/1698059955/3edc8af363b692abdbacba2141a26691a03ea52e/source-s3/4.1.2/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:3edc8af363b692abdbacba2141a26691a03ea52e)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-s3 test report (commit [`0550b77564`](https://github.com/airbytehq/airbyte/pull/31383/commits/0550b775649372774bf5ccd65c4e9f807507263c)) - ❌

⏲️  Total pipeline duration: 03mn29s 

| Step                                                     | Result   |
|:---------------------------------------------------------|:---------|
| Build source-s3 docker image for platform(s) linux/amd64 | ❌       |
| Code format checks                                       | ✅       |
| Validate metadata for source-s3                          | ✅       |
| Connector version semver check                           | ✅       |
| Connector version increment check                        | ✅       |
| QA checks                                                | ✅       |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/tolik0_source-s3_wrong-bucket-config-error/1698060875/0550b775649372774bf5ccd65c4e9f807507263c/source-s3/4.1.2/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:0550b775649372774bf5ccd65c4e9f807507263c)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```

**Comment by airbyte-oss-build-runner** [comment]
##  source-s3 test report (commit [`fe9d87a2fd`](https://github.com/airbytehq/airbyte/pull/31383/commits/fe9d87a2fd646ce86cc980ca5c97353820deb8fd)) - ✅

⏲️  Total pipeline duration: 17mn56s 

| Step                                                     | Result   |
|:---------------------------------------------------------|:---------|
| Build source-s3 docker image for platform(s) linux/amd64 | ✅       |
| Unit tests                                               | ✅       |
| Acceptance tests                                         | ✅       |
| Check our base image is used                             | ✅       |
| Code format checks                                       | ✅       |
| Validate metadata for source-s3                          | ✅       |
| Connector version semver check                           | ✅       |
| Connector version increment check                        | ✅       |
| QA checks                                                | ✅       |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/tolik0_source-s3_wrong-bucket-config-error/1698177254/fe9d87a2fd646ce86cc980ca5c97353820deb8fd/source-s3/4.1.2/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:fe9d87a2fd646ce86cc980ca5c97353820deb8fd)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-s3 test
```
